### ISSUE TITLE
Type deduction in foo.py module does not respect foo.pyi file

### ISSUE THREAD

**Issue by yaxum62** [microsoft/pyright#1129]
**Describe the bug**
Type checking in foo.py module does not respect foo.pyi stub file.

For example given module `foo.py`

```
class Vector3d:
  def __init__(self, x:int = 0, y:int = 0, z:int = 0):
    self.x=x
    self.y=y
    self.z=z

class Vector2d(Vector3d):
  pass
```

and stub file `foo.pyi`

```
class Vector2d:
  def __init__(self, x:int = 0, y:int = 0):
    ...
```

Pyright does not report error for `v = Vector2d(20, 30, 40)` in foo.py, while same code in other module yields expected error `Expected 2 positional arguments`.

**VS Code extension or command-line**
vscode extension v1.1.82

**Comment by erictraut** [comment]
I'm not sure I understand your bug report.

Type checkers always prefer a type stub ("*.pyi") file over an implementation file ("*.py") when resolving an import. So if you have two files `a.pyi` and `a.py` in the same directory and a third file (b.py) contains the statement `from . import a`, the type checker will use the contents of `a.pyi` for type information.

Are you expecting that when the type checker analyzes `a.py` that it will somehow use the contents of `a.pyi` in the analysis? That's not how type stubs work.

**Comment by yaxum62** [comment]
Sorry I pressed enter too quick, updated details in description.

**Comment by erictraut** [comment]
Yeah, that's not how type stubs work. The information in a type stub is used when the module is imported. In your example, there is no import of `foo`. When the type checker is analyzing the code in `foo.py`, it analyzes all of the code within that file; it doesn't somehow "merge" the information in a type stub by the same name.

**Comment by erictraut** [comment]
A type stub should accurately reflect the interface for the corresponding module. In your example above, the Vector2 constructor described in the ".pyi" file doesn't match the signature of the constructor in the implementation. So there's a bug in either your implementation or your type stub.

**Comment by yaxum62** [comment]
Thanks for checking! I search around but did not found something mentioned stub file only take effect during module importing, nor something requires stub must be exactly same with implementation. An obvious case that stub is narrower than actual implementation is when using @typing.override in stub file.

I understand that currently pyright only check stub file during importing, but do you think it is possible to apply same check for module itself? I can make this a feature request if you think it better.

**Comment by yaxum62** [comment]
I made a better use case for this issue:

```
# foo.py
def func(x):
  print(x)

func([]) # this passed type check where stub file restricted func to int and str only.

# foo.pyi
def func(x:int)...
def func(x:str)...
```

**Comment by erictraut** [comment]
No, that feature wouldn't make sense. As I said, a type stub should always accurately reflect the interface provided by the implementation. It shouldn't be "narrower" or differ in other ways. If the implementation of Vector2's constructor is intended to accept only two parameters, then it should be implemented as such:

```python
class Vector2d(Vector3d):
  def __init__(self, x:int = 0, y:int = 0):
      super().__init__(x, y)
```

**Comment by erictraut** [comment]
For your second example, an overload wouldn't make sense here. Overloads are useful when the return type differs based on the parameter types.

If you really want `func` to accept only `int` and `str` parameters, then you would implement it as such:

```python
def func(x: Union[int, str]):
    print(x)
```

I think you are confused about the purpose of a type stub file. They are meant to declare the interface to a module that will be consumed (imported) by another module — especially in situations where the implementation is in another language like C++.

If you are writing a module that is internal to your project and you are implementing it in Python, you should simply include the type information inline within the code. There's no need for a type stub file in this case. You can use overloads, generics, etc. within your implementation.

**Comment by yaxum62** [comment]
For the example, I made it too simple for understanding but lost its justification. An more practical case could be something like `open(file: str, mode: Literal['rb']) -> IO[bytes]` and `open(file: str, mode: Literal['r']) -> IO[str]`. You know, abuse of function override.

I understand that for most cases I can just add type hint to code itself. But I have my problem - generating type hint for a home-made decorator similar to `@dataclass`. Generated stub files seems a way out, the standard does not against it, but I ran into this issue.

I created this issue without clear scenario because I tried to avoid making it just for my problem. If you think it is not a generic problem that worth to address, feel free to close it.
